### ISSUE TITLE
yggdrasil.sock already exists after write /dev/net/tun: no buffer space available crash

### ISSUE THREAD

**Issue by whoizit** [yggdrasil-network/yggdrasil-go#319]
yggdrasil 0.3.2
Already two times I getting this error. Yggdrasil is down for some reason and can't restart. Is it a distro specific problem, or is it a problem Ygg? Void Linux distribution (runit) have supervisor and trying to restart it, but it can't, cause yggdrasil.sock already exists.
```
2019/01/30 18:13:47 WARNING: /var/run/yggdrasil.sock already exists and may be in use by another process
2019/01/30 18:13:47 Admin socket failed to listen: listen unix /var/run/yggdrasil.sock: bind: address already in use
```
```
$ sudo cat /etc/sv/yggdrasil/run 
#!/bin/sh
if [ -f /etc/yggdrasil.conf ]; then
	exec /usr/bin/yggdrasil --useconffile /etc/yggdrasil.conf
else
	exec /usr/bin/yggdrasil --autoconf
fi
```

**Comment by neilalexander** [comment]
I think `yggdrasil.sock` would remain there if Yggdrasil crashed for some reason, in which case if the Yggdrasil process is definitely not running then it's safe to delete. It might be worth keeping some logs to see what's going wrong.

**Comment by whoizit** [comment]
```
panic: write /dev/net/tun: no buffer space available

goroutine 5 [running]:
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*tunAdapter).write(0xc0001122c8, 0x0, 0x0)
	/builddir/yggdrasil-go-0.3.2/src/yggdrasil/tun.go:175 +0x8f3
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*tunAdapter).start.func2(0xc0001122c8)
	/builddir/yggdrasil-go-0.3.2/src/yggdrasil/tun.go:60 +0x2f
created by github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*tunAdapter).start
	/builddir/yggdrasil-go-0.3.2/src/yggdrasil/tun.go:60 +0x113
```

**Comment by neilalexander** [comment]
Hm, okay, so is this 0.3.2 from `master` branch or is it a `develop` build? Are you running in TUN mode or TAP mode?

**Comment by whoizit** [comment]
it's 0.3.2 release (https://github.com/void-linux/void-packages/blob/master/srcpkgs/yggdrasil/template)

absolutelly default config, only add 3 finland nodes

**Comment by whoizit** [comment]
It's TUN

**Comment by whoizit** [comment]
my load it's my HTTP(S) proxy for telegram and browser [202:9877:2815:cd91:336:2a16:bfd9:7258]:8118

**Comment by neilalexander** [comment]
Does it help at all if you reduce the `IfMTU` setting to something smaller? Try standard ones like 1500, 9000 and then perhaps keep increasing until you hit the problem again.

**Comment by whoizit** [comment]
I have:
```
net.core.rmem_default = 212992
net.core.rmem_max = 212992
net.core.wmem_default = 212992
net.core.wmem_max = 212992
```

**Comment by neilalexander** [comment]
Did lowering `IfMTU` in your Yggdrasil configuration help at all?

**Comment by Arceliar** [comment]
@whoizit from that "while" it sounded like maybe you had more to say?

In any case, this sounds very similar to the issues we were seeing with TAP devices. Lowering MTU seemed to help in those cases too. I wonder if the buffer allocated by the OS for the tun/tap isn't big enough, and the tap just hits it easier for whatever reason. I'm puzzled by why we don't see this elsewhere (e.g. iperf3 tests). I wonder if maybe something differs from distro to distro...

I'm almost certain that the error is an `ENOBUFS`, but I have no idea why a tuntap write would cause that for large MTU and not for small MTU, unless smaller MTU just slows things down enough that other things drain the buffer. In which case, I don't know what the right thing to do is (loop until we either finish or get an error that isn't `ENOBUFS` or `EAGAIN`, or just drop the packet because we're apparently congested).

**Comment by whoizit** [comment]
while means I need a time for tests. I can't more to say, cause I don't speak english.

TUN means TUN:
```
7: tun0: <POINTOPOINT,UP,LOWER_UP> mtu 65535 qdisc pfifo_fast state UNKNOWN group default qlen 500
    link/none 
    inet6 200:c9:d951:8509:78b:6010:d760:ae53/7 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::9908:89ec:1058:97eb/64 scope link stable-privacy 
       valid_lft forever preferred_lft forever
```

**Comment by whoizit** [comment]
can't reproduce anymore with mtu 65535, while

**Comment by whoizit** [comment]
rly hard to catch, can reproduce not often

**Comment by whoizit** [comment]
```
panic: write /dev/net/tun: no buffer space available

goroutine 40 [running]:
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*tunAdapter).write(0xc0000b0e88, 0x0, 0x0)
	/builddir/yggdrasil-go-c5cc280556450a79c5dbbf7c1dc0a39cacbfda91/src/yggdrasil/tun.go:201 +0x8e5
github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*tunAdapter).start.func2(0xc0000b0e88)
	/builddir/yggdrasil-go-c5cc280556450a79c5dbbf7c1dc0a39cacbfda91/src/yggdrasil/tun.go:86 +0x2f
created by github.com/yggdrasil-network/yggdrasil-go/src/yggdrasil.(*tunAdapter).start
	/builddir/yggdrasil-go-c5cc280556450a79c5dbbf7c1dc0a39cacbfda91/src/yggdrasil/tun.go:86 +0x2da

```

**Comment by neilalexander** [comment]
Does this still happen with version 0.3.3?

**Comment by whoizit** [comment]
for `IfMTU: 1500` seems like works, not tested yet for 65535

**Comment by whoizit** [comment]
hard to reproduce

**Comment by Arceliar** [comment]
@neilalexander We have better logging now, so we could print a warning message whenever the tun has a write error, but maybe get rid of the `panic`. If nothing else, a `panic` in this goroutine seems to prevent the `defer` statements in others from being run, so that keeps the admin socket from closing cleanly, which is annoying. If it's this hard to reproduce, then it must not be a very common issue--if it's not a problem anymore after a packet or two, then it's probably not worth killing the application, and in any case we can't check if it's still a problem for later packets if we're `panic`ing the first time it happens.

**Comment by whoizit** [comment]
can't reproduce

**Comment by whoizit** [comment]
again `Admin socket failed to listen: listen unix /var/run/yggdrasil.sock: bind: address already in use`
I'll just add `rm /var/run/yggdrasil.sock` to the startup script and forget about it.

**Comment by neilalexander** [comment]
So we have a patch in `develop` at the moment (and should be merged into v0.3.4 soon) which should solve the problem with the admin socket not being deleted. Are you able to test for a couple of days with `develop`?

**Comment by neilalexander** [comment]
Closing this for now as it *should* be fixed in v0.3.4/v0.3.5 by #356 and #358. Please reopen the issue if this keeps happening.
